Implement auto-tiling

MarianoGnu opened the pull request:
This PR adds auto-tiling to TileMap and TileSet classes, it's not finished, but having some feedback while working on it is worth, please have in mind i'll squash commits every time i make some changes so merge this to your forks or to master is a bad idea.

How it works:
- A single tile should have a texture and asign the region of the texture where all subtiles exists. The subtiles can be sepparated by a fixed amount of pixels
- The BitmaskMode property on the tileset allows to define the size of the matrix 
- The tile is marked as autotile on the inspector
- When editing a TileSet a SideControl on the CanvasEditor and a button with the text "Autotiles" at the bottom tools will appear. these two are used to edit autotiles properties.
- Tiles with the IsAutotile property will be listed on the side list picker. When selecting one, it's autotile properties can be edited under the list, properties are: spacing and size.
- After setting up size and spacing click on the "Autotiles" button at the bottom pannel to open the sub editors. The editor will display the region of the texture used in the tile and a grid representing the subdivision according to the size and spacing configured on the autotile properties
- The "Icon" edit mode is a sinple picker to select the subtile wich represents the autotile on the TileMap Editor, also this subtile will be used when drawing a tile wich bitmask configuration is not been setled yet.
- The "Bitmask" editor allows to specify what pars of the tile belong to the same terrain, this is what the BitmaskMode property is for, depending on this proeprty, every subtile could have 4 or 9 bits to be enabled/disabled
- Collision Oclusion and Navigation editors are not functional yet, and might need some design desitions to define the UX
- Besides these editors there are two more features that should be implemented:
1. Subtile Priority: when editing bitmask, you can set the same bitmask to many subtiles, when this happens godot will pick one of them randomly when drawing, setting a priority should change how often one of them is used in comparison to all the others.
2. Afinity-Friendship: this property has not a defined name yet: when editing you might want a terrain to behabe to other tiles like if it would be of the same time, this prevents creating borders beatween, for example, two kinds of watter.

**Changes on TileSet and TileSetEditor**
- [x] New Properties for every tile: is_autotile, autotile_spacing, autotile_size, autotile_icon_coord, autotile_coord_flag
- [x] Allow to store collision oclusion and navigation shapes for every subtile
- [x] Properties to set Priority and Afinity
----------
Create editors for editing Autotile specific properties:
- [x] Icon
- [x] Bitmask
- [x] Collision
- [x] Oclusion
- [x] Navigation
- [x] Priority
- [x] Afinity/Friendship

**Changes on TileMap and TileMapEditor**
- [x] functions to interact with bitmask (make_bitmask_dirty, update_dirty_bitmask, update_cell_bitmask, update_area_bitmask)
- [x] show auto-tile icon instead of complete texture region
- [x] save auto-tile data along with cells data
- [x] update bitmask data while drawing

**Known Issues**
- ~Current TileMap Save/Load Method breaks backward compatibility (if i change this then this feature could be merged to 2.1.x)~
- ~Maps can't be saved, the process is corrupted and causes a SIGABRT when saving.~
- ~The property "IsAutotile" can be changed but it's not saved/loaded~
- ~When selecting a TileMap or TileSet to be edited, the side-editor is squashed on the left of the CanvasItemEditor~
- ~When Editing a TileSet, it is necesary to change selection of the autotile to refresh AutoTile and show the editors properly~

MarianoGnu committed:
Implement auto-tiling

MarianoGnu commented:
Update: Added toolbars for every edition mode, probably not all tools thought, and some tools have not explanatory icons

MarianoGnu commented:
Update
Fixed:
- ~The property "IsAutotile" can be changed but it's not saved/loaded~
- ~When selecting a TileMap or TileSet to be edited, the side-editor is squashed on the left of the CanvasItemEditor~

blurymind commented:
This is absolutely fantastic! :D Really hope it gets in godot 3

MarianoGnu commented:
Update:
Collision editor and collisions themselves are working

damarindra commented:
Proposal: I'm using 4k monitor, and the autotile editor is really small, can you add zoom option?

MarianoGnu commented:
Yeah, i was leaving it for the last step, since there's few time for merging new features before Godot3.0 is released i thought i could leave that out of this PR and submit an issue for that

MarianoGnu commented:
Update, fix recent conflics

MarianoGnu commented:
Update occlusion editor and API are fully working

MarianoGnu commented:
Navigation is working now, we are almost there, with 90% of the PR done only the other 90% is left.

MarianoGnu commented:
Update: *PR is 99% done*
- Fixed merge conflicts
- Renamed all new functions on TileSet for readability and consistency
- Made BitmaskMode property per tile instead of per TileSet (this means you can use both modes on the same TileSet)
- Added a virtual function to let the user decide what tiles bind each other `_is_tile_binded(drawn_id, neighbor_id)` this replaces the afinity/friendship property proposal (script must be attached to TileSet and use tool option
- Added a virtual function to let the user choose a subtile to use on the tileset `func _forward_subtile_selection(autotile_id, bitmask, tilemap, tile_location)` (script must be attached to TileSet and use tool option
-Added priority editor and API to set/get/save/load this data

TODO:
- Implement bitmask copy/overryde/clear buttons
- Implement zoom in/out (optional?)

MarianoGnu commented:
Update:
- Implemented copy/override/clear bitmask buttons
- Implemented Zoom In/Out (there you go @damarindra :smile: )

MarianoGnu commented:
Only missing step is fixing the text format, @akien-mga where was the link to the automation tool? i tryed to install the QtCreator plugin but it broke the format instead of fixing it :(

vnen commented:
@MarianoGnu you can install the [pre-commit hook](http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#pre-commit-hook) to fix the formatting when you commit.

MarianoGnu commented:
Thx a lot @vnen i already aplyed the patches.
@akien-mga, @reduz, @godotengine/user-interface this PR is ready for review

hoelzl commented:
MarianoGnu: That's a great new feature, I'm looking forward to trying it out.

Just a minor quibble: Should the function `_is_tile_binded` not perhaps be named `_is_tile_bound`?

MarianoGnu commented:
whatever makes more sense to users, @godotengine/documentation  team should know

bojidar-bg commented:
"binded" is not correct, since "bind" is an irregular verb having "bound" as past tense.

MarianoGnu commented:
Thx for clarification, function renamed to `_is_tile_bound`

MarianoGnu commented:
Demo by @damarindra 
https://www.facebook.com/damar.indera/videos/10210628918500786/

nhydock commented:
Can you post an example of an auto-tile compatible asset to know what the spec looks like.  When I hear auto-tile, I think of the kind that RPG Maker uses.

MarianoGnu commented:
the difference betwheen rpg maker and this implementation is that RPGMaker expects an specific layout on the texture and tile size, this one is much more flexible and workflow is based on Tiled Terrains: https://www.youtube.com/watch?v=SIYNuXdDClU

nhydock commented:
Hmm, the advantage I find with using the RPG Maker specific layout is the fact that there's a bunch of those assets that exist already, and forcing a spec makes it easier to share the assets too that'll be compatible with other people's projects.  I guess if this solution is flexible enough that I can just create a resource definition that maps to the rpg maker layout and then I can just swap the texture, it should be easy enough to make use of assets I already have.

MarianoGnu commented:
there are copy and paste bitmask layout buttons, so if you have a lot of siminar textures you have to tet it up just once (copy/paste buttons are missing for shapes) besides, adding a duplicate button should be very easy, that way the only thing you have to do is duplicate the tile data and change the texture or texture region.

blurymind commented:
It would be awesome if autotile resources are easy to share between godot users :D 

For gods sake please get this merged before the feature freeze. It is too awesome to miss

blurymind commented:
@MarianoGnu can you please add an example file somewhere- so we can study what the best practices are to set this up

blurymind commented:
Suggestion -if a tile that is available in the tileset is not set to 'Autotile', still list it in the tiles list on the left- but make the bitmask options and tab un-available for it.
That way we can use the editor on both normal and auto tiles
![normaltileautotilemixing](https://user-images.githubusercontent.com/6495061/33189455-ae56eee4-d09a-11e7-803e-744bd74c2f64.JPG)

Also it would be nice to be able to actually save the edited tileset with the autotile settings and bitmask to a scene file, which can be shared. I guess the only way to do it at the moment is to put the tilemap node that has it in a scene file?
